onnxruntime/core/platform/env.h

```diff
@@ -111,17 +111,37 @@ class Env {
       MappedMemoryPtr& mapped_memory) const = 0;
 
 #ifdef _WIN32
+  /// \brief Returns true if the directory exists.
+  virtual bool FolderExists(const std::wstring& path) const = 0;
+  /// \brief Recursively creates the directory, if it doesn't exist.
+  virtual common::Status CreateFolder(const std::wstring& path) const = 0;
+  // Recursively deletes the directory and its contents.
+  // Note: This function is not thread safe!
+  virtual common::Status DeleteFolder(const std::wstring& path) const = 0;
   //Mainly for use with protobuf library
   virtual common::Status FileOpenRd(const std::wstring& path, /*out*/ int& fd) const = 0;
   //Mainly for use with protobuf library
   virtual common::Status FileOpenWr(const std::wstring& path, /*out*/ int& fd) const = 0;
 #endif
+  /// \brief Returns true if the directory exists.
+  virtual bool FolderExists(const std::string& path) const = 0;
+  /// \brief Recursively creates the directory, if it doesn't exist.
+  virtual common::Status CreateFolder(const std::string& path) const = 0;
+  // Recursively deletes the directory and its contents.
+  // Note: This function is not thread safe!
+  virtual common::Status DeleteFolder(const std::string& path) const = 0;
   //Mainly for use with protobuf library
   virtual common::Status FileOpenRd(const std::string& path, /*out*/ int& fd) const = 0;
   //Mainly for use with protobuf library
   virtual common::Status FileOpenWr(const std::string& path, /*out*/ int& fd) const = 0;
   //Mainly for use with protobuf library
   virtual common::Status FileClose(int fd) const = 0;
+
+  /** Gets the canonical form of a file path (symlinks resolved). */
+  virtual common::Status GetCanonicalPath(
+      const std::basic_string<ORTCHAR_T>& path,
```

Comment:
When is it needed? Why symlink doesn't fit into your needs?

Reply:
Symlinks are not desirable when computing correct relative paths.

E.g., given A = /a/b/symlink_to_a, B = /a/b/c, the pure relative path computation from A to B is ../c but we want b/c.

Reply:
When need to compute the correct relative path?

Reply:
The relative path is used when loading checkpoint TensorProtos with external data ([https://github.com/microsoft/onnxruntime/blob/4b2c8e884e258af619ce13240820f8b299082212/orttraining/orttraining/core/framework/checkpointing.cc#L272](https://github.com/microsoft/onnxruntime/blob/4b2c8e884e258af619ce13240820f8b299082212/orttraining/orttraining/core/framework/checkpointing.cc#L272)).

---
In reply to: [393882658](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r393882658) [](ancestors = 393882658)

Reply:
I think it is not needed, at least the ONNX spec didn't say much on this. 
Let's say, if you have model with path of:
/home/edgchen/model.onnx,
and the real path is:
/data/edgchen/model.onnx
If the external data is at /data/weights
Then if you set the relative path to ../weights , if we compute it with '/home/edgchen/model.onnx', then it can't be found. 
If the spec didn't say people must use real path here, then we should avoid such conversion ?

Reply:
We talked offline. For the case where TensorProtos are loaded from checkpoints, resolving symlinks for the relative path computation is the desired behavior. However, it may not be the case in general.

---
In reply to: [395779207](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r395779207) [](ancestors = 395779207)